### PR TITLE
Feature 2072 tc rmw winds

### LINKED ISSUE #2072: Enhance TC-RMW to compute tangential and radial winds.
## Describe the Enhancement ##
When preparing training material for the MET-TC tools and also applying it for the unified physics testing and evaluation project, @KathrynNewman found that the TC-RWM tool does not actually perform the steps that the documentation claims it does. Namely, it does not compute and report tangential and radial winds. Looking at the tc_rmw.cc source code, I agree with her.

While the [wind_ne_to_ra()](https://github.com/dtcenter/MET/blob/af2ce1307fbdaf22b5a36727665c7a562352d10b/met/src/tools/tc_utils/tc_rmw/tc_rmw.cc#L630) function looks like logic for winds, that function is never actually being called! Instead [process_fields()](https://github.com/dtcenter/MET/blob/af2ce1307fbdaf22b5a36727665c7a562352d10b/met/src/tools/tc_utils/tc_rmw/tc_rmw.cc#L710) just loops over the requested fields but includes no special logic for handling U and V.

Reading this, [tc_rmw chapter](https://met.readthedocs.io/en/develop/Users_Guide/tc-rmw.html) of the User's Guide, I agree that the functionally does not match the documentation. This needs more work.

This task is enhance the TC-RMW tool to compute and report tangential and radial winds. Listing @KathrynNewman as an assignee so that she can advise on and help test the development.

### Time Estimate ###
2 days

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
No sub-issues needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2793451 - based on guidance on Kathryn

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
None.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by davidalbo**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
New configuration file options were added to control conversion of winds:
```
//
// Optionally convert u/v winds to tangential/radial winds
//
compute_tangential_and_radial_winds = TRUE;
u_wind_field_name = "UGRD";
v_wind_field_name = "VGRD";
tangential_velocity_field_name = "VT";
tangential_velocity_long_field_name = "Tangential Velocity";
radial_velocity_field_name = "VR";
radial_velocity_long_field_name = "Radial Velocity";
```
- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>
The tangential and radial velocity are new netCDF output variables.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>    The unit tests (2 of them) for this app were modified to include the new configuration parameters.  The tests ran successfully.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>   Possibly run the unit tests and then look at the output netCDF files where the new output fields can be seen.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**   Not sure.  Is there other documentation besides the user guide?

- [x] Do these changes include sufficient testing updates? **[Yes or No]**  Not sure.  Are there other tests besides the unit tests?

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>  The output of the unit tests has new netCDF variables for radial and tangential winds.

- [ ] Please complete this pull request review by **[May 23, 2022]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by JohnHalleyGotway** [comment]
@davidalbo, I took a look at this PR and realize that the checks didn't look right because you branched from develop a long time ago. I'm going to do the following to try to get the checks to run:
```
git checkout develop; git pull # To get the latest version of develop
git checkout feature_2072_tc_rmw_winds; git pull # To get the latest from your feature branch
git merge develop # To merge develop into your feature branch
git push # To push the result back up to GitHub
```
I think that'll trigger the full set of tests to run... but will let you know.

File: src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    else if (varname == _conf->v_wind_field_name.string()) {
+      _vIndexMap[varlevel] = i_var;
+    }
+  }
+  // test for consistency
+  if (_uIndexMap.size() != _vIndexMap.size()) {
+    mlog << Error << " error uneven number of ugrid/vgrid, no wind conversion\n";
+    mlog << Error << _conf->u_wind_field_name.string() << " had " << _uIndexMap.size() << " inputs\n";
+    mlog << Error << _conf->v_wind_field_name.string() << " had " << _vIndexMap.size() << " inputs\n";
+    _computeWinds = false;
```
Comment (JohnHalleyGotway) [inline src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc:100]: Everywhere else in MET, all ERROR messages are followed by a call to exit(1) to return bad status.
Recommend either exiting here, or change the "Error" messages to "Warning" instead.

File: src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  }
+  map<string,int>::const_iterator iu, iv;
+  for (iu=_uIndexMap.begin(), iv=_vIndexMap.begin(); iu!=_uIndexMap.end(); ++iu, ++iv) {
+    if (iu->first != iv->first) {
+      mlog << Error << "error ordering of ugrid/vgrid input levels not the same, not implemented, no wind conversions\n";
+      mlog << Error << "    "  << iu->first  << " " << iv->first << "\n";
+      _computeWinds = false;
```
Comment (JohnHalleyGotway) [inline src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc:107]: Everywhere else in MET, all ERROR messages are followed by a call to exit(1) to return bad status.
Recommend either exiting here, or change the "Error" messages to "Warning" instead.

File: src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    variable_long_names[_conf->tangential_velocity_field_name] = _conf->tangential_velocity_long_field_name.string();
+    variable_units[_conf->tangential_velocity_field_name] = _unitsU;
+    variable_levels[_conf->radial_velocity_field_name] = variable_levels[_conf->u_wind_field_name.string()];
+    variable_long_names[_conf->radial_velocity_field_name] = _conf->radial_velocity_long_field_name.string();
+    variable_units[_conf->radial_velocity_field_name] = _unitsU;
+  }
+  else {
+    if (!_foundUInInput) {
+      mlog << Error << "\nTCWRMW_WindConverter::checkInputs() -> "
+	   << "field not found in input" << _conf->u_wind_field_name << "\n";
```
Comment (JohnHalleyGotway) [inline src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc]: Everywhere else in MET, all ERROR messages are followed by a call to exit(1) to return bad status.
Recommend either exiting here, or change the "Error" messages to "Warning" instead.

```suggestion
	   << "field not found in input \"" << _conf->u_wind_field_name << "\"\n";
```

File: src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    if (!_foundVInInput) {
+      mlog << Error << "\nTCWRMW_WindConverter::checkInputs() -> "
+	   << "field not found in input" << _conf->v_wind_field_name << "\n";
```
Comment (JohnHalleyGotway) [inline src/tools/tc_utils/tc_rmw/tc_rmw_wind_converter.cc]: Everywhere else in MET, all ERROR messages are followed by a call to exit(1) to return bad status.
Recommend either exiting here, or change the "Error" messages to "Warning" instead.

```suggestion
	   << "field not found in input \"" << _conf->v_wind_field_name << "\"\n";
```

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
- Please see individual comments about Error vs Warning messages. Either changes the errors to warnings or keep them as error and exit after printing the error message.
- Throughout MET we're using spaces rather than tabs for indents but keeping that consistent seems like a uphill battle/lost cause. Just wanted to mention it.

- After merging the latest from the develop branch into this feature branch... and rerunning bootstrap on seneca to regenerate the Makefile.in files... the GHA testing workflow ran. As expected, it flagged differences in 2 TCRMW output files, and none else:

tc_rmw_hwrf_gonzalo.nc and tc_rmw_pressure_lev_out.nc now contain 2 new output variables:
```
> 	double VR(range, azimuth, pressure, track_point) ;
> 		VR:long_name = "Radial Velocity" ;
> 		VR:units = "m/s" ;
> 		VR:_FillValue = -9999. ;
> 	double VT(range, azimuth, pressure, track_point) ;
> 		VT:long_name = "Tangential Velocity" ;
> 		VT:units = "m/s" ;
> 		VT:_FillValue = -9999. ;
```

To test further, I compiled this feature branch locally on seneca. I reconfigured/reran to confirm that...
- setting ```compute_tangential_and_radial_winds = FALSE;``` disables them.
- setting ```u_wind_field_name``` to a bad string triggers an error.
- changing the string names has the desired effect in the output.
- if UGRD and VGRD are NOT requested in the input, then the wind derivation errors out, as it should.

@davidalbo please address the error vs warning issue and then "Re-Request" my review when you're all done.

**Comment by davidalbo** [comment]
Thanks for the thorough feedback @JohnHalleyGotway .  The spaces versus tabs thing...I understand the rationale, but.. I'm editing with emacs which puts in the tabs, making it very hard to comply.  Maybe there is a way to turn that off, not sure.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

Thanks for switching those errors to warnings. As we discussed via email, I tweaked the wording to reference u/v winds. And I consolidated multiple calls to the logger into a single call over multiple lines. Both technically work, but we generally do the latter elsewhere.

Now that the PR is approved, please go ahead and click the "squash and merge" button to merge the changes into the develop branch. That's the job of the submitter... unless they ask that the reviewer do it for them.

Thanks Dave!

**Pull request merged**